pkg/ast/spl/spl.peg

```diff
@@ -40,6 +40,130 @@ func ConvertCustomDateTimeFormatToEpochMs(dateStr string) (int64, error) {
     return t.UnixMilli(), nil
 }
 
+
+func applySnap(snap string, t time.Time) (time.Time, error) {
+    sec := t.Second()
+    min := t.Minute()
+    hour := t.Hour()
+    day := t.Day()
+    week := t.Weekday()
+    mon := t.Month()
+    year := t.Year()
+    
+    if snap[0] != 'w' {
+        tunit, err := strconv.Atoi(snap)
+        if err != nil {
+            return t, fmt.Errorf("Error while converting the snap: %v to integer, err: %v", snap, err)
+        }
+
+        switch utils.TimeUnit(tunit) {   
+            case utils.TMSecond:
+                return time.Date(year, mon, day, hour, min, sec, 0, time.Local), nil
+            case utils.TMMinute:
+                return time.Date(year, mon, day, hour, min, 0, 0, time.Local), nil
+            case utils.TMHour:
+                return time.Date(year, mon, day, hour, 0, 0, 0, time.Local), nil
+            case utils.TMDay:
+                return time.Date(year, mon, day, 0, 0, 0, 0, time.Local), nil
+            case utils.TMWeek:
+                diff := week - time.Sunday
+                return time.Date(year, mon, day-int(diff), 0, 0, 0, 0, time.Local), nil
+            case utils.TMMonth:
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMQuarter:
+                if mon >= time.October {
+                    mon = time.October
+                } else if mon >= time.July {
+                    mon = time.July
+                } else if mon >= time.April {
+                    mon = time.April
+                } else {
+                    mon = time.January
+                }
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMYear:
+                return time.Date(year, 1, 1, 0, 0, 0, 0, time.Local), nil
+            default:
+                return t, fmt.Errorf("Unsupported time unit for relative timestamp: %v", tunit)
+        }
+    } else {
+        if len(snap) != 2 {
+            return t, fmt.Errorf("Error for special week snap, should be of type w[0-7] got: %v", snap)
+        }
+        weeknum := int(snap[1] - '0')
+        if weeknum == 7 {
+            weeknum = 0
+        }
+        diff := int(week)-weeknum
+        if diff < 0 {
+            diff += 7
+        }
+        return time.Date(year, mon, day, 0, 0, 0, 0, time.Local).AddDate(0, 0, -diff), nil
+    }
+}
+
+
+func applyOffset(num int64, unit utils.TimeUnit, t time.Time) (time.Time, error) {
+
+    durNum := time.Duration(num)
```

Comment:
I think renaming `num` will help make this function more clear. Perhaps `nanoseconds` would be a better choice. What do you think?

Reply:
num is actually the number of times you want to apply the offset so basicall the offset number.

Reply:
Ah ok, that makes sense. Then why do you have `durNum := time.Duration(num)`? I see later you do `t.Add(durNum * time.Second)` but couldn't you do `t.Add(num * time.Second)` instead and then there's no point in having `durNum`? Or will that not work?